### PR TITLE
Fix failing rollback - #127

### PR DESCRIPTION
**Pull request by jan-gerling**
* Safeguard the transaction rollback operation

### CODE REVIEW

File: data-collection/src/main/java/refactoringml/db/Database.java [hunk 1]
```diff
@@ -74,8 +78,29 @@ public long findAllStableCommits(long projectId, int level) {
...
+	public void rollback() throws SQLException {
+		//this session object itself should never be null, thus we don't check for it
+		//nothing to do in this case, recovering the session or transaction is to much effort and the db takes care of a failed transaction
+		if(!session.isOpen()) {
+			log.error("Encountered is closed session during rollback. Doing nothing.");
+			return;
+		}
+
+		if(!session.isConnected()){
+			Connection connection =	sf.getSessionFactoryOptions().getServiceRegistry().
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/db/Database.java]: that's magic! I didn't know that before, but I suppose you tried and it worked?

File: data-collection/src/main/java/refactoringml/db/Database.java [hunk 2]
```diff
@@ -74,8 +78,29 @@ public long findAllStableCommits(long projectId, int level) {
...
+					getService(ConnectionProvider.class).getConnection();
+			session.reconnect(connection);
+		}
+
+		//standard case for a rollback
+		if(session.isConnected() && session.getTransaction() != null) {
+			session.getTransaction().rollback();
+		}
+		//other cases:
+		//1. not connected to the DB : we could raise an error here, because something is probably wrong with the db
```
Comment (mauricioaniche) [inline data-collection/src/main/java/refactoringml/db/Database.java]: We could indeed throw an exception here, and catch it in the `App`. But, pragmatically, that won't change anything. Maybe not letting the exception to bubble up is also good, as we have nothing to do if the exception happens anyways.

(Good thing is that the `exceptionsCount` int will increase, so later we will be able to count how many commits we missed.)

File: data-collection/src/main/java/refactoringml/db/Database.java (hunk 1 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/db/Database.java]: It works under some conditions

File: data-collection/src/main/java/refactoringml/db/Database.java (hunk 2 above)
Comment (jan-gerling) [reply data-collection/src/main/java/refactoringml/db/Database.java]: > (Good thing is that the `exceptionsCount` int will increase, so later we will be able to count how many commits we missed.)

That is a good remark, I will throw an exception here.

**Pull request merged**